Add parsing Kerberos PAC-ResourceGroups

ankor2023 opened the pull request:
Domain local AD-groups are placed in the PAC-ResourceGroups block.
Parse this block to have information about the user's membership in these groups.

rousskov reviewed (commented):
Thank you for contributing this feature. If possible, please adjust PR description to explicitly disclose how/whether these changes affect existing helper deployments and to briefly note how the new feature can be used (or what it can be used for).

IMO, proposed code violates several basic engineering principles, but I assume it is mimicking existing Kerberos parsing code in Squid and that it has been proven to "work" for some use cases, so I am not going to vote against this PR. One of those principles is reviewability; FWIW, I am unable to properly review this kind of code and, hence, will not be requesting any changes.

I suggest declaring the whole helper as unsupported by the Project, so that there is less pressure on the Project to fix this helper when somebody bothers to test this code thoroughly and publish a dozen vulnerability reports against it. Such a declaration would also simplify accepting unreviewed helper code. CC: @yadij, @kinkie.

rousskov commented:
OK to test

ankor2023 commented:
@rousskov 
Alex, thank you for the comments.

> If possible, please adjust PR description to explicitly disclose how/whether these changes affect existing helper deployments and to briefly note how the new feature can be used (or what it can be used for).
> 
I have adjusted the description.

> IMO, proposed code violates several basic engineering principles, but I assume it is mimicking existing Kerberos parsing code in Squid and that it has been proven to "work" for some use cases, so I am not going to vote against this PR. One of those principles is reviewability; FWIW, I am unable to properly review this kind of code and, hence, will not be requesting any changes.
> 
I do not know how to improve the reviewwability of the code. The new functions are based on the  previous code which parse GroupIds and ExtraSids. I can try to rewrite the code of the negotiate_kerberos_pac.cc but I don't know the mentioned "basic engineering principles".

> I suggest declaring the whole helper as unsupported by the Project, so that there is less pressure on the Project to fix this helper when somebody bothers to test this code thoroughly and publish a dozen vulnerability reports against it. Such a declaration would also simplify accepting unreviewed helper code. CC: @yadij, @kinkie.
>
I think the Kerberos helper is too important for the squid-users, and it would be nice to keep its support.

I also added the definitions of the new functions to negotiate_kerberos.h to fix the github CI build problem.

ankor2023 commented:
> I think the Kerberos helper is too important for the squid-users, and it would be nice to keep its support.

Since the code of the Kerberos helper has been unchanged for many years, I think the current version fully meets the needs of the squid-community.
So we can reject my PR, especially taking into account the lack of resources of the Squid-team.

yadij commented:
@ankor2023; please do not be intimidated by @rousskov opinions. IME he has a string tendency to say these same things about any code he does not personally see people complaining about or wanting constant changes to.

@rousskov; this is one of the primary helpers needed/used by admin managing a network with ActiveDirectory authentication. Lack of feedback is more an indication of how it is working for that group without causing them trouble. AFAIK Markus Moeller is still supporting the code.

rousskov commented:
> I have adjusted the description.

Thank you very much! Assuming  your commit [8c0b194](https://github.com/squid-cache/squid/pull/1597/commits/8c0b1940820b037216d1c049d17de843fd4797b9) fixes CI tests, the ball is now fully on the Squid Project side.
 
> > Alex: IMO, proposed code violates several basic engineering principles, but I assume it is mimicking existing Kerberos parsing code in Squid and that it has been proven to "work" for some use cases, so I am not going to vote against this PR. One of those principles is reviewability; FWIW, I am unable to properly review this kind of code and, hence, will not be requesting any changes.
 
> Andrey: I do not know how to improve the reviewwability of the code.

AFAICT, it would require a serious rewrite, and I am _not_ asking for that, especially since such a rewrite is likely to create tensions with the existing Kerberos parsing code.

> The new functions are based on the previous code which parse GroupIds and ExtraSids. I can try to rewrite the code of the negotiate_kerberos_pac.cc but I don't know the mentioned "basic engineering principles".

Glad we are on the same page. I do not recommend a rewrite in this case.

> > Alex: I suggest declaring the whole helper as unsupported by the Project, so that there is less pressure on the Project to fix this helper when somebody bothers to test this code thoroughly and publish a dozen vulnerability reports against it. Such a declaration would also simplify accepting unreviewed helper code.

> Andrey: I think the Kerberos helper is too important for the squid-users, and it would be nice to keep its support.

I agree, and my recommendation does not contradict the above assertions. My recommendation takes the current state of code and Project resources into account: There is not much we can do here in the short term to solve the problems that need to be solved. IMO, it is much better to be honest about serious problems than to pretend they do not exist; while awareness of problem existence may intimidate, that negative/unwanted side effect is better than the false sense of security.

> Amos: this is one of the primary helpers needed/used by admin managing a network with ActiveDirectory authentication.

My recommendation does not contradict the above fact.

> Amos: AFAIK Markus Moeller is still supporting the code.

IMO, if Markus is indeed supporting this code, he should be reviewing this PR. If you can solicit his review, please do so; I am not sure whether the following mention reaches Markus. CC: @huaraz.

> Andrey: we can reject my PR, especially taking into account the lack of resources of the Squid-team.

I (still) do _not_ recommend rejection at this time because the proposed feature is obviously useful, and it does not make the helper worse (there are no known _new_ flaws in the proposed code). While I cannot properly review the details, I do support accepting this PR, especially if others can review this code (or at least share my opinion that this code should be accepted). I am approving this PR without clearing it for merging so that @huaraz, @yadij, and @kinkie have plenty of time to properly review this PR.

rousskov reviewed (approved):
The code should be formatted before clearing it for merging.

huaraz commented:
I can have a look at the PR over the next week or so.

Markus

huaraz commented:
The change looks fine i.e. is in line with the way I extract SIDs and resource groups are relevant for some cases of forest trusts.  I am only surprised nobody mentioned this use case earlier.  

https://learn.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2003/cc772808(v=ws.10)?redirectedfrom=MSDN

Markus

rousskov commented:
@huaraz, thank you for reviewing this code!

@yadij, @ankor2023, will you address the minor problems identified in my earlier [review](https://github.com/squid-cache/squid/pull/1597#pullrequestreview-1746687813) or should we merge this PR "as is"?

ankor2023 commented:
@huaraz: Hello, Markus, 

The code uses RPC_UNICODE_STRING data structures:
```
typedef struct {
    uint16_t length;
    uint16_t maxlength;
    uint32_t pointer;
} RPC_UNICODE_STRING;

```
It corresponds to the Microsoft [article](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/94a16bb6-c610-4cb9-8db6-26f15f560061). I can't figure out where the **uint32_t pointer** points. Can I use it to access the actual data inside the KERB_VALIDATION_INFO structure without having to iterate through the following KERB_VALIDATION_INFO elements? 
If it's not too much trouble, could you clarify a little?

ankor2023 commented:
@rousskov, @yadij :  
Hello, Alex and Amos,

The PR has **s-waiting-for-author** tag.
Do I need to take any action?

Kind regards,
      Ankor.

yadij commented:
> The PR has **s-waiting-for-author** tag. Do I need to take any action on my part?

I added it initially for @rousskov question. Those changes have now been done. If you think all the changes that need to be done are completed please request the re-review and switch to S-waiting-for-reviewer.

rousskov commented:
> Alex: The code should be formatted before clearing it for merging.

> Andrey: The PR has **s-waiting-for-author** tag. Do I need to take any action on my part?

I see two actions that are not required, but could speed things up:

1. If getting your [question](https://github.com/squid-cache/squid/pull/1597#issuecomment-1837352251) answered is not required to merge this PR, please say so explicitly. It was not clear to me whether we should wait for that answer...

2. As I said, it would be nice to format new/changed sources before merging, but you are _not_ required to do that. A core developer that clears this PR for merging can do that (by running ./scripts/source-maintenance.sh and then removing out-of-scope changes). FWIW, the minimal/automated formatting fixes are visible as a diff at the end of test-suite/test-sources.sh test [log](https://github.com/squid-cache/squid/actions/runs/7074336950/job/19255916694?pr=1597).

Others may request other actions/changes, of course, but I am not aware of any other outstanding requests (or blocking reviews).

yadij commented on src/auth/negotiate/kerberos/negotiate_kerberos_pac.cc:
```diff
@@ -362,6 +367,114 @@ getextrasids(char *ad_groups, uint32_t ExtraSids, uint32_t SidCount)
+
+        uint32_t nauth = get4byt();
+
+        size_t length = 1+1+6+nauth*4;
```
Please document what characters the magic numbers are reserving space for.

ankor2023 replied:
I commented this string in a new file version:
// length = revision[1byte]+nsub[1byte]+idauth[6bytes]+nauth*sauth[4bytes]
size_t length = 1+1+6+nauth*4;

yadij commented on src/auth/negotiate/kerberos/negotiate_kerberos_pac.cc:
```diff
@@ -362,6 +367,114 @@ getextrasids(char *ad_groups, uint32_t ExtraSids, uint32_t SidCount)
+
+        ag=(char *)xcalloc((length+1)*sizeof(char),1);
+        // the first byte is a length of the SID
+        ag[0] = (char) length;
```
The binary value of length is 4 bytes long. Please validate that the stored value will fit into this 1-byte space without truncation.

ankor2023 replied:
I changed the store space to 4-bytes in a new file version.

yadij commented on src/auth/negotiate/kerberos/negotiate_kerberos_pac.cc:
```diff
@@ -362,6 +367,114 @@ getextrasids(char *ad_groups, uint32_t ExtraSids, uint32_t SidCount)
+        for ( l=0; l<(int)nauth; l++ ) {
+            uint32_t sauth;
+            sauth = get4byt();
+            snprintf((char *)&dli[strlen(dli)],sizeof(dli)-strlen(dli),"-%u",sauth);
```
Please handle `snprintf()` error return codes.

ankor2023 replied:
I added assert() of the returned value in a new file version.

yadij commented on src/auth/negotiate/kerberos/negotiate_kerberos_pac.cc:
```diff
@@ -362,6 +367,114 @@ getextrasids(char *ad_groups, uint32_t ExtraSids, uint32_t SidCount)
+}
+
+static char *
+get_resource_groups(char *ad_groups, char *resource_group_domain_sid, uint32_t ResourceGroupIds, uint32_t ResourceGroupCount) {
```
```suggestion
get_resource_groups(char *ad_groups, char *resource_group_domain_sid, uint32_t ResourceGroupIds, uint32_t ResourceGroupCount)
{
```

yadij reviewed (changes requested):
Ignoring most of the formatting for now. There are some potential error cases that seem important to handle.

ankor2023 commented:
> Ignoring most of the formatting for now. There are some potential error cases that seem important to handle.


@yadij  
Hello, Amos, 

Thank you very much for the review!
I will check it.

ankor2023 commented on src/auth/negotiate/kerberos/negotiate_kerberos_pac.cc:
```diff
@@ -49,6 +49,11 @@ static unsigned char *p;
 extern int
 check_k5_err(krb5_context context, const char *function, krb5_error_code code);
 
+static char *
+get_resource_group_domain_sid(uint32_t ResourceGroupDomainSid);
+static char *
+get_resource_groups(char *ad_groups, char *resource_group_domain_sid, uint32_t ResourceGroupIds, uint32_t ResourceGroupCount);
+
```
Previously, I did not include the definitions in the code, but it did not pass github automatic checks. 
So I added these lines to negotiate_kerberos.h and moved them to negotiate_kerberos_pac.cc after the Alex's review.

rousskov replied:
@ankor2023, I suspect that you have misinterpreted the meaning of an earlier build [error](https://github.com/squid-cache/squid/actions/runs/6952672981/job/18934998337#step:4:5009) for the original PR commit 8450f61ff4fb2a5015ead043b11a31d744262618:

    src/auth/negotiate/kerberos/negotiate_kerberos_pac.cc:411:1: error: no previous declaration for
    ‘char* get_resource_groups(char*, char*, uint32_t, uint32_t)’ [-Werror=missing-declarations]


In that commit, get_resource_groups() was a "global" function -- something other translation units can call directly by name. The compiler was telling us that a _global_  function needs a declaration. 

```
missing-declarations
  Warn if a global function is defined without a previous
  declaration.  Do so even if the definition itself provides a
  prototype.  Use this option to detect global functions that are not
  declared in header files.
```

Today, get_resource_groups() is a local function. When a local static function is not used prior to its definition, it does not need a declaration. Naturally, Squid code is not consistent with avoiding needless local function declarations, but it is best to avoid them.

ankor2023 replied:
@rousskov : Alex, thank you for the clarifications!

ankor2023 commented:
> Ignoring most of the formatting for now. There are some potential error cases that seem important to handle.

@yadij : Hello, Amos,

I posted a new version of the file, where I fixed your comments, renamed some variables to more appropriate ones and added some checks.
Could you review it if you have time?

ankor2023 commented on src/auth/negotiate/kerberos/negotiate_kerberos_pac.cc:
```diff
@@ -362,6 +367,114 @@ getextrasids(char *ad_groups, uint32_t ExtraSids, uint32_t SidCount)
+        bpos = bpos + 1; /*nsub*/
+        idauth = get6byt_be();
+
+        snprintf(dli,sizeof(dli),"S-%d-%lu",rev,(long unsigned int)idauth);
```
I added assert() of the returned value in a new file version.

ankor2023 commented:
> > @yadij requested a review from rousskov on December 5, 2023 02:19
> 
> My position has not changed since [#1597 (review)](https://github.com/squid-cache/squid/pull/1597#pullrequestreview-1744593768)


> I suggest declaring the whole helper as unsupported by the Project, so that there is less pressure on the Project to fix this helper when somebody bothers to test this code thoroughly and publish a dozen vulnerability reports against it. Such a declaration would also simplify accepting unreviewed helper code.

@rousskov  : Hello, Alex,

Could you please explain how it will affect the squid users? Maybe accepting unreviewed code is not good for the project, especially considering the importance of this helper?
Maybe there are other developers in your core team who could help Markus Moeller support the code?

kinkie commented:
> There are currently three core developers: @kinkie, @yadij, and @rousskov. I do not speak for others, but all core developers are (or, IMHO, should be) reading comments on this PR and, I bet, the three of us disagree what "support the code" really means in this context, whether this particular code is currently "supported" (by anybody), etc. We are also, IMHO, overloaded, so it is not really a question of finding a core developer to assign a new responsibility to; it is a question of what are we willing to sacrifice in exchange for that new assignment.

I agree with what @rousskov wrote, I would like to highlight however how wonderful it would be if
more people, expressing their abilities through successful contributions, could join the group of core developers :)

Regarding whether this code should be part of the core distribution or not: our shared goal is to provide the most reliable, powerful, and easy to manage tool we can given our resources.

Maybe a possible way forward is to resurrect a contrib/ subtree, which would highlight how the core developers cannot provide guarantees over the quality of what is contained there, while trying to maximise convenience for downstream administrators, and possibly empowering some contributors to tune their skills there - in the end that's how I approached the project, too

rousskov commented:
> Regarding whether this code should be part of the core distribution or not

Please note that nobody has suggested to remove this helper from Squid distribution. I have suggested to declare the helper unsupported, but that is a very different suggestion.

> Maybe a possible way forward is to resurrect a contrib/ subtree

IMO, it is best to avoid placing source code based on its support status (and moving code when its support status changes). A lot of code, especially helper code, requires certain build infrastructure/context/surroundings. We do not want to duplicate that, maintaining, say, two helper collections, in two different source tree locations. There are other, cheaper/lightweight/flexible ways to identify unsupported code (but we better define "supported" before finalizing our indexing approach!).

ankor2023 commented:
> There are currently three core developers: @kinkie, @yadij, and @rousskov.

@kinkie , @yadij , @rousskov 
You are three programmers and you support and develop such a complex system.
I respect you very much!
I'm sure you know how it would be better for the project to declare the helper unsupported, or leave it as it is.

Thank you!

yadij commented on src/auth/negotiate/kerberos/negotiate_kerberos_pac.cc:
```diff
@@ -362,6 +365,154 @@ getextrasids(char *ad_groups, uint32_t ExtraSids, uint32_t SidCount)
+        }
+
+        // length = revision[1byte]+nsub[1byte]+idauth[6bytes]+nauth*sauth[4bytes]
+        size_t length = 1 + 1 + 6 + nauth*4;
+
+        sid=(char *)xcalloc((length + sizeof(uint32_t))*sizeof(char), 1); // +sizeof(uint32_t) for storing length of a SID
```
FYI, I have omitted the `*sizeof(char)` from first parameter of these calls to `xcalloc()` because the second parameter is where the array object size is supposed to be passed.
 ... and `xcalloc(N, 1)` is already just shorthand for `xcalloc(N, sizeof(char))`

ankor2023 commented on src/auth/negotiate/kerberos/negotiate_kerberos_pac.cc:
```diff
@@ -362,6 +365,154 @@ getextrasids(char *ad_groups, uint32_t ExtraSids, uint32_t SidCount)
+        debug((char *) "%s| %s: INFO: Found %d Resource Group rids\n", LogTime(), PROGRAM, ResourceGroupCount);
+
+        // make a group template which begins with the ResourceGroupDomainID
+        length = group_domain_sid_len + 4;  // +4 for a rid concatenation
+        st = (char *)xcalloc(length*sizeof(char), 1);
```
@yadij 
Hello, Amos,

Thank you so much for such a detailed and valuable review.
I will try to take into account all your comments and correct the code in the near future.

@yadij @kinkie @rousskov : 
I wish you a happy New Year!

ankor2023 commented:
@yadij 
Hello, Amos,

Thank you so much for such a detailed and valuable review.
I will try to take into account all your comments and correct the code in the near future.

@yadij @kinkie @rousskov : 
I wish you a happy New Year!

ankor2023 commented:
@yadij : 
Hello, Amos,

I've corrected your comments in the code. Let's wait a couple of weeks to test it.

yadij reviewed (approved):
I am not seeing anything major now.

Have added a few code-style polish and compiler warning fixes. No behaviour changes expected.

yadij commented:
Seeing as there has been no logic updates, nor negative feedback in the past month I am clearing this for merge to Squid-7 as an experimental feature addition.
